File: packages/typescript/src/recorder.ts
import type { TokenomicsEvent } from "./types.js";

export interface Sink { emit(event: TokenomicsEvent): void | Promise<void> }

export class MemorySink implements Sink {
  readonly events: TokenomicsEvent[] = [];
  emit(event: TokenomicsEvent): void { this.events.push(event); }
}

export class MultiSink implements Sink {
  constructor(readonly sinks: readonly Sink[]) {}
  async emit(event: TokenomicsEvent): Promise<void> {
    for (const sink of this.sinks) await sink.emit(event);
  }
}

export class Recorder {
  constructor(readonly sink: Sink) {}
  async record(event: TokenomicsEvent): Promise<TokenomicsEvent> {
    await this.sink.emit(event);
    return event;
  }
}

File: packages/python/src/tokenomics/recorder.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Protocol

from .models import TokenomicsEvent


class Sink(Protocol):
    def emit(self, event: TokenomicsEvent) -> None: ...


@dataclass
class MemorySink:
    events: list[TokenomicsEvent] = field(default_factory=list)

    def emit(self, event: TokenomicsEvent) -> None:
        self.events.append(event)


class MultiSink:
    def __init__(self, *sinks: Sink):
        self.sinks = tuple(sinks)

    def emit(self, event: TokenomicsEvent) -> None:
        for sink in self.sinks:
            sink.emit(event)


class Recorder:
    """Minimal fan-out recorder. Applications own instrumentation timing."""

    def __init__(self, sink: Sink):
        self.sink = sink

    def record(self, event: TokenomicsEvent) -> TokenomicsEvent:
        self.sink.emit(event)
        return event
